types/register.d.ts

```diff
@@ -1,13 +1,14 @@
 import { FastifyPluginOptions, FastifyPluginCallback, FastifyPluginAsync } from './plugin'
 import { LogLevel } from './logger'
+import { FastifyInstance } from './instance'
 
 interface RegisterOptions {
   prefix?: string;
   logLevel?: LogLevel;
   logSerializers?: Record<string, (value: any) => string>;
 }
 
-export type FastifyRegisterOptions<Options> = (RegisterOptions & Options) | (() => RegisterOptions & Options)
+export type FastifyRegisterOptions<Options> = (RegisterOptions & Options) | ((instance: FastifyInstance) => RegisterOptions & Options)
 
 /**
  * FastifyRegister
```

Comment:
@sinclairzx81  I think I found a missing part in the [FastifyRegister interface](https://github.com/fastify/fastify/pull/3398/files#diff-3ab92153aecdba9f45a29d5fa78392abd148e8e9841436d84e5e712d0542fc94R18-R31) when using the `instance.register` will complain when trying to use a TypeProvider. I believe this is the code missing

```ts
import { RawServerBase } from "./utils";
import { FastifyTypeProvider, FastifyTypeProviderDefault } from "./type-provider";
```
```ts
export interface FastifyRegister<T = void> {
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginCallback<Options, Server, TypeProvider>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginAsync<Options, Server, TypeProvider>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginCallback<Options, Server, TypeProvider> | FastifyPluginAsync<Options, Server, TypeProvider> | Promise<{ default: FastifyPluginCallback<Options, Server, TypeProvider> }> | Promise<{ default: FastifyPluginAsync<Options, Server, TypeProvider> }>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
}
```

Reply:
@rickerp Nice! Are you are to test with a couple of community plug ins (like `fastify-swagger`) ? The following are a couple of tests which should help verify the change. (should be able to drop the following in the https://github.com/fastify/fastify/blob/next/test/types/type-provider.test-d.ts file)

```typescript
import fastifySwagger from 'fastify-swagger'

// --------------------------------------------------------------------
// Utility Type (pending object interface refactor)
// --------------------------------------------------------------------

interface FastifyInstanceWithTypeProvider<TypeProvider extends FastifyTypeProvider> extends FastifyInstance<
  RawServerDefault,
  RawRequestDefaultExpression<RawServerDefault>,
  RawReplyDefaultExpression<RawServerDefault>,
  FastifyLoggerInstance,
  TypeProvider
> {}

const fastify = Fastify() // Fastify will use FastifyTypeProviderDefault as default

// --------------------------------------------------------------------
// Local Plugins
// --------------------------------------------------------------------

async function anyTypePlugin(instance: any, options: any) { }

async function typeProviderPlugin(instance: FastifyInstanceWithTypeProvider<TypeBoxProvider>, options: { xyz: string }) { }

// --------------------------------------------------------------------
// Tests
// --------------------------------------------------------------------

// Register should infer for any plugin arguments.
fastify.register(anyTypePlugin, { xyz: '123' }) 

// Register should infer for varying Type Providers (i.e. TypeBoxProvider !== FastifyTypeProviderDefault)
fastify.register(typeProviderPlugin, { xyz: '123' }) 

// Register should infer from existing plugins that use a typed plugin signature (but are non-TypeProvider aware)
fastify.register(fastifySwagger, {}) 
```
If all checks out, you should submit a PR. 

Good work!

Reply:
No ,I tested wit my own plugins (like "routers")

Reply:
Could you create an issue to keep tracking this issue?